lib/src/metta/runner/mod.rs

```diff
@@ -18,10 +18,10 @@ const EXEC_SYMBOL : Atom = sym!("!");
 
 #[derive(Debug, Clone)]
 pub struct Metta {
-    space: Shared<GroundingSpace>,
+    space: Shared<SpaceBox>,
```

Comment:
This is the most confusing question of the review to me. Using `SpaceBox` everywhere instead of `GroundingSpace` effectively leads to the using `Rc<RefCell<Box<dyn MutSpace>>>` everywhere. Which seems unnecessary complicated because we could have just `Rc<RefCell<dyn MutSpace>>` instead. In addition `Shared` is actually used mostly with `Space` instances (not only with them, but most of the usages keep spaces). So `Shared<SpaceBoxed>` raises a desire to replace it by a single type. I don't have good solution yet, but I think we need to think about removing this complexity.

Reply:
`Rc` and `RefCell` both act like pointers in Rust, but in fact they're not guaranteed to be heap-allocated.  This manifests as the `Rc` or `RefCell` taking the `Sized` marker trait status of whatever they contain.  In the case of a dyn object the size isn't knowable at compile time.  `Box`, on the other hand, is `Sized` (128 bits on a 64 bit system, to be exact).  The good news is that the compiler eliminates the unnecessary pointer-chase and collapses it down.

As far as `Shared<T>`, I am not a fan of this approach.  It has already caused some annoyances - although nothing serious yet - but we might find ourselves in a place where we have painted ourselves into a corner.  For example, https://github.com/trueagi-io/hyperon-experimental/issues/241 . I suspect there will be some more pain to come when we begin thinking about multi-threading, and we might use that opportunity to rethink resource management.

For the time being, however, I think the inner Box<dyn Trait> is the best solution.

Reply:
A few updates after a night of sleep.
1.) What I said about Rc not being able to guarantee a heap allocation was actually wrong.  It turns out it can do what a Box does in that regard.

2.) However, removing the inner box doesn't let us totally eliminate the extra pointer-indirection, because the allocation metadata associated with the allocation (Rc or Box) can't be reconstructed from a raw pointer for a dynamic type.

3.) Therefore, by moving from an inner-box to an outer box, we can make the Rust side a tiny bit better, but it's a wash on the C side. (We will never notice this change, performance-wise, and the compiler will take care of it either way)

4.) I took your suggestion to unify the Shared<T> with the SpaceBox, and created the DynSpace type.  It makes the code a tiny bit cleaner, and allows some impls to be deprecated.

5.) My objection to embedded RefCells in an interface is still very strong.  We can revisit this when it's time to tackle multi-threading.

Changes in 61c3c9d